list() return all drives

nytamin opened the pull request:
Closes #24 


This PR changes how list() fetches drives. It'll now return all drives with status info, not only the ones with OK status.

This is a **BREAKING CHANGE**:
Return type of list() has changed, from a key-value with strings to a key-value with objects containing statuses and path.
Return type of find() has changed, from a string to an object containing statuses and path.


Notes: 
* I have tested this to work locally and the unit tests pass locally
* The README mentions linting and formatting using VS Code. But no linter, linting-rules or formatting-rules are defined, so I haven't done this.

nytamin committed:
Change how list() fetches drives. It'll now return all drives with status info, not only the ones with OK status.

BREAKING CHANGES

Return type of list() has changed, from a string to an object containing statuses and path.
Return type of find() has changed, from a string to an object containing statuses and path.

larrybahr commented on index.js:
```diff
@@ -127,21 +131,21 @@ let windowsNetworkDrive = {
 	 * @function list
 	 * @public
-	 * @returns {Promise<Object>} - Object keys are drive letters, values are the network path
+	 * @returns {Promise<Object>} - Object keys are drive letters, values are { ok: boolean, path: string }
```
What about `statusString`

nytamin replied:
You're correct, of course. I'll fix!

larrybahr commented on index.d.ts:
```diff
@@ -1,9 +1,26 @@
 declare module 'windows-network-drive' {
   /** Finds if a path is already mounted and returns all drive letters that point to that exact path. */
-  function find(drivePath: string): Promise<string[]>;
+  function find(drivePath: string): Promise<{
```
Since this in now a more complex object, it could make sense to have a single interface that both `find` and `list` return. Maybe something like
``` typescript
interface driveInfo {
    /** Status of the network drive. A falsy value might indicate connectivity issues */
    status: boolean

    /** Drive letter */
    driveLetter: string

    /** Path to the network drive */
    path: string

    /**
     * Status string, describing the status of the network drive.
     * This is a textual value depending on the local Windows system language.
     */
    statusMessage: string
}
```

nytamin replied:
Good idea!

larrybahr commented on index.js:
```diff
@@ -127,21 +131,21 @@ let windowsNetworkDrive = {
-	list: function list()
-	{
-		let listPromise;
+	 list: function list()
```
Looks like some extra white space. VSCode has a default formatter that I use. This is a small project so using the built in one just made more sense at the time.

nytamin replied:
I was able to figure out which formatter is the default (I'm using Prettier locally, it had to be disabled for the default to take over). I also took the liberty of updating the README to clarify which formatter to use.

larrybahr commented on index.js:
```diff
@@ -151,50 +155,71 @@ let windowsNetworkDrive = {
+				const lines = `${result.stdout}`
+					.replace(/^(-+)$/gm, '') // remove the "-----------------------------"-line
+					.split('\n')
+					.map((l) => l.replace(/[\r\n]/g, '')) // Trim line endings
```
IMHO single letter vars lead to confusion in the long term. Could we change `l` to `line`?

When possible I prefer self documenting code. In this case there is a built in method that does exactly what you want `line.trim()` so you do not need the regex.

nytamin replied:
sure, good points!

larrybahr commented on index.js:
```diff
@@ -151,50 +155,71 @@ let windowsNetworkDrive = {
+					.replace(/^(-+)$/gm, '') // remove the "-----------------------------"-line
+					.split('\n')
+					.map((l) => l.replace(/[\r\n]/g, '')) // Trim line endings
+					.filter(Boolean) // Remove empty lines
```
You would not need the comment if you did something like `line => line.length !== 0`

nytamin replied:
I'm inclined to leave this as-is, tbh.
`.filter(Boolean)` is the safest way of removing falsy values and I think that the verbosity in the comments for these operations are important to follow what's happening.

larrybahr replied:
Fair enough :+1:

larrybahr commented on index.js:
```diff
@@ -151,50 +155,71 @@ let windowsNetworkDrive = {
+				// The command completed successfully.
+
+
+				const lines = `${result.stdout}`
```
Why is this wrapped in a template literal? Could you just use `result.stdout`?

nytamin replied:
Out of habit, I suppose.
It's a way to guard for the case where stdout is not a string (ie a Buffer).

larrybahr commented on index.js:
```diff
@@ -151,50 +155,71 @@ let windowsNetworkDrive = {
+						const statusString = m[1].trim();
+						const driveLetter = m[2].trim().toUpperCase();
+						const path = m[3].trim();
+						// const network = m[4].trim();
```
Why have commented out code? If this is useful info you should just return it.

nytamin replied:
The only value I've seen in this field is "Microsoft Windows Network", I don't really see the use for it, but if you think it's useful I can include it.

larrybahr commented on index.js:
```diff
@@ -151,50 +155,71 @@ let windowsNetworkDrive = {
+
+
+				// Fix an issue where the table rows are split in multiple lines, because of a long path:
+				for (let i = 0; i< lines.length; i++) {
```
`i` => `index`

larrybahr commented on index.js:
```diff
@@ -151,50 +155,71 @@ let windowsNetworkDrive = {
-						let driveLetter = currentPath.substring(0, colonIndex);
-						let drivePath = currentPath.substring(colonIndex + 1);
+				const drivePaths = {}
+				for (let i = 0; i< lines.length; i++) {
```
`i` => `index`

larrybahr commented on index.js:
```diff
@@ -151,50 +155,71 @@ let windowsNetworkDrive = {
+				for (let i = 0; i< lines.length; i++) {
+					const line = lines[i];
+
+					const m = line.match(/^(.+) +(\w): +([^ ]+) +(.+)$/);
```
`m` => `matches`

larrybahr commented on test/test.js:
```diff
@@ -491,6 +491,39 @@ describe('windows-network-drive', function ()
+
+
+					const driveLetters = Object.keys(driveList)
+					for (let i = 0; i < driveLetters.length; i++) {
```
`i` => `index`

larrybahr commented on test/test.js:
```diff
@@ -491,6 +491,39 @@ describe('windows-network-drive', function ()
+					}
+
+					const firstResult = driveList[driveLetters[0]];
+					if (!firstResult) {
```
I would do an explicit type check here `'object' === typeof firstResult`

larrybahr commented on test/test.js:
```diff
@@ -491,6 +491,39 @@ describe('windows-network-drive', function ()
+						throw (new Error("Bad result data. driveList = " + JSON.stringify(driveList, null, '\t')));
+					}
+
+					if (firstResult.status !== Boolean(firstResult.status)) {
```
What about `typeof firstResult.status !== 'boolean'`

larrybahr commented on test/test.js:
```diff
@@ -491,6 +491,39 @@ describe('windows-network-drive', function ()
+						}
+					}
+
+					const firstResult = driveList[driveLetters[0]];
```
Why only check the first result? Since you are already looping over the drive letters you could do these checks in the same loop.

nytamin replied:
Since there is a check just above that asserts the length to be 1, I didn't see the point.

larrybahr replied:
I am on a small screen so I may have missed it, but I do not see an assertion that forces it to be 1 and only 1. I do see an assertion that will make sure we have at least 1 result.

larrybahr reviewed (changes requested):
Thank you for the great improvements! You did a good job writing new test cases and I enjoyed looking over your work.

I have a couple small suggestions that I would like to hear your thoughts on.

larrybahr commented:
Also feel free to add yourself as a contributor https://github.com/larrybahr/windows-network-drive/blob/master/package.json#L11

nytamin committed:
chore: reformat, using VSCode's  "Typescript and JavaScript Language Features" formatter

nytamin committed:
fix: unify return types of list() and find()

nytamin committed:
chore: update README with note on which formatter to use

nytamin committed:
chore: clarify line ending trim

nytamin committed:
chore: clarifications after PR feedback

nytamin committed:
chore: add contributor

nytamin commented:
@larrybahr Thanks for the thorough review!
I've addressed and fixed most of your comments, please have another look and let me know if there's anything left you feel particularly strongly about.

larrybahr reviewed (commented):
Thank you for making the changes. I think it looks great overall. I just want to confirm the tests are good before I accept (see note). Everything else looks good!

larrybahr commented:
FYI, I added automated tests via github actions that should run if/when you push to this PR. :crossed_fingers: they run.

larrybahr commented on README.md:
```diff
@@ -166,7 +166,10 @@ $ npm test
 
 ## Contributing
 
```
The list and find docs in the ReadMe.md need to be updated

larrybahr reviewed (changes requested):
More minor changes:
* The README.md needs updated (see notes)
* The examples in `example/write-a-file-to-a-network-drive/index.js` also need to be updated since the interface is changing

nytamin committed:
chore: update docs

nytamin committed:
chore: update example

nytamin commented:
good points! I've updated the docs and example

larrybahr commented:
Thank you for your hard work/many revisions :handshake: . Everything looks great!
I will merge this and work on releasing it.

larrybahr commented:
V4.0.0 is live on NPM with this change
